spam/03_spam_data_slicing_tutorial.py

```diff
@@ -0,0 +1,499 @@
+
+# %%
+from snorkel.slicing import slice_dataframe
+
+short_link_df = slice_dataframe(df_valid, short_link)
+short_link_df[["text", "label"]]
+
+# %% [markdown]
+# ## 2. Train a discriminative model
+#
+# To start, we'll initialize a discriminative model using our [`MultitaskClassifier`](https://snorkel.readthedocs.io/en/redux/packages/_autosummary/classification/snorkel.classification.MultitaskClassifier.html#snorkel.classification.MultitaskClassifier).
+# We'll assume that you are familiar with Snorkel's multitask model — if not, we'd recommend you check out our [Multitask Tutorial](https://github.com/snorkel-team/snorkel-tutorials/blob/master/multitask/multitask_tutorial.ipynb).
+#
+# In this section, we're ignoring slice information for modeling purposes; slices are used solely for monitoring fine-grained performance.
+
+# %% [markdown]
+# ### Featurize Data
+#
+# First, we'll featurize the data—as you saw in the introductory Spam tutorial, we can extract simple bag of words features and store them as numpy arrays.
+
+# %%
+import torch
+from sklearn.feature_extraction.text import CountVectorizer
+
+vectorizer = CountVectorizer(ngram_range=(1, 1))
+
+
+def df_to_torch_features(vectorizer, df, fit_train=False):
+    words = [row.text for i, row in df.iterrows()]
+
+    if fit_train:
+        feats = vectorizer.fit_transform(words)
+    else:
+        feats = vectorizer.transform(words)
+    X = feats.todense()
+    Y = df["label"].values
+    return X, Y
+
+
+# %%
+X_train, Y_train = df_to_torch_features(vectorizer, df_train, fit_train=True)
+X_valid, Y_valid = df_to_torch_features(vectorizer, df_valid, fit_train=False)
+X_test, Y_test = df_to_torch_features(vectorizer, df_test, fit_train=False)
+
+# %% [markdown]
+# ### Create DataLoaders
+#
+# Next, we'll use the extracted Tensors to initialize a [`DictDataLoader`](https://snorkel.readthedocs.io/en/redux/packages/_autosummary/classification/snorkel.classification.DictDataLoader.html) — as a quick recap, this is a Snorkel-specific class that inherits from the common PyTorch class and supports multiple data fields in the `X_dict` and labels in the `Y_dict`.
+#
+# In this task, we'd like to store the bag-of-words `bow_features` in our `X_dict`, and we have one set of labels (for now) correpsonding to the `spam_task`.
+
+# %%
+from snorkel.classification.data import DictDataset, DictDataLoader
+
+BATCH_SIZE = 32
+
+
+def create_dict_dataloader(X, Y, split, **kwargs):
+    """Create a DictDataLoader for bag-of-words features."""
+    ds = DictDataset(
+        name="spam_dataset",
+        split=split,
+        X_dict={"bow_features": torch.FloatTensor(X)},
+        Y_dict={"spam_task": torch.LongTensor(Y)},
+    )
+    return DictDataLoader(ds, **kwargs)
```

Comment:
I would also lean towards this being in a util file- it's separate, cluttering up an already packed tutorial.  Also I really don't think users should have to reason about *tasks* at all in this tutorial, e.g. `spam_task`